pulse2percept/models/_axon_map.pyx

```diff
@@ -0,0 +1,142 @@
+
+    cdef double[:] xprime = np.empty_like(rho)
+    cdef double[:] yprime = np.empty_like(rho)
+    rho_min = c_min(rho)
+    for idx in range(len(rho)):
+        tmp_rho = rho[idx]
+        tmp_phi = phi0 + b * c_pow(tmp_rho - rho_min, c)
+        xprime[idx] = tmp_rho * c_cos(deg2rad * tmp_phi)
+        yprime[idx] = tmp_rho * c_sin(deg2rad * tmp_phi)
+    return np.asarray(xprime), np.asarray(yprime)
+
+
+cdef np.intp_t argmin_segment(double[:, :] bundles, double x, double y):
+    cdef double dist2, min_dist2
+    cdef np.intp_t seg, min_seg, n_seg
+
+    min_dist2 = 1e12
+    n_seg = bundles.shape[0]
+    for seg in range(n_seg):
+        dist2 = c_pow(bundles[seg, 0] - x, 2) + c_pow(bundles[seg, 1] - y, 2)
+        if dist2 < min_dist2:
+            min_dist2 = dist2
+            min_seg = seg
+    return min_seg
+
+
+cpdef fast_axon_contribution(double[:, :] bundle, double[:] xy, double lmbd):
+    cdef np.intp_t p, c, argmin, n_seg
+    cdef double dist2
+    cdef double[:, :] contrib
+
+    # Find the segment that is closest to the soma `xy`:
+    argmin = argmin_segment(bundle, xy[0], xy[1])
+
+    # Add the exact location of the soma:
+    bundle[argmin + 1, 0] = xy[0]
+    bundle[argmin + 1, 1] = xy[1]
+
+    # For every axon segment, calculate distance from soma by summing up the
+    # individual distances between neighboring axon segments
+    # (by "walking along the axon"):
+    n_seg = argmin + 1
+    contrib = np.zeros((n_seg, 3))
+    dist2 = 0
+    c = 0
+    for p in range(argmin, -1, -1):
+        dist2 += (c_pow(bundle[p, 0] - bundle[p + 1, 0], 2) +
+                  c_pow(bundle[p, 1] - bundle[p + 1, 1], 2))
+        contrib[c, 0] = bundle[p, 0]
+        contrib[c, 1] = bundle[p, 1]
+        contrib[c, 2] = c_exp(-dist2 / (2.0 * c_pow(lmbd, 2)))
+        c += 1
+    return np.asarray(contrib)
+
+
+cpdef fast_finds_closest_axons(double[:, :] bundles, double[:] xret,
+                               double[:] yret):
+    cdef np.intp_t[:] closest_seg = np.empty(len(xret), dtype=int)
+    cdef np.intp_t n_xy, n_seg
+    n_xy = len(xret)
+    n_seg = bundles.shape[0]
+    for pos in range(n_xy):
+        closest_seg[pos] = argmin_segment(bundles, xret[pos], yret[pos])
+    return np.asarray(closest_seg)
+
+
+cpdef fast_axon_map(double[:] stim, double[:] xel, double[:] yel,
+                    double[:, ::1] axon, double rho, double th):
+    cdef np.intp_t idx, n_stim, n_ax
+    cdef double bright, gauss
+    n_stim = len(stim)
+    n_ax = axon.shape[0]
+    cdef double[:] act = np.zeros(n_ax)
+    with nogil:
+        for i_stim in range(n_stim):
+            for i_ax in range(n_ax):
+                r2 = c_pow(axon[i_ax, 0] - xel[i_stim], 2)
+                r2 += c_pow(axon[i_ax, 1] - yel[i_stim], 2)
+                gauss = c_exp(-r2 / (2.0 * c_pow(rho, 2)))
+                act[i_ax] += stim[i_stim] * axon[i_ax, 2] * gauss
+    bright = c_max(act)
+    if bright < th:
+        bright = 0
+    return bright
```

Comment:
I noticed that all the functions here are called `fast_XXX`. Maybe we can drop that prefix?